AutoCollection/HeartBeat.ts

```diff
@@ -0,0 +1,138 @@
+import os = require("os");
+import TelemetryClient = require("../Library/TelemetryClient");
+import Constants = require("../Declarations/Constants");
+import Util = require("../Library/Util");
+import Config = require("../Library/Config");
+import Context = require("../Library/Context");
+
+class HeartBeat {
+
+    public static INSTANCE: HeartBeat;
+
+    private static AIMS_URI = "http://169.254.169.254/metadata/instance/compute";
+    private static AIMS_API_VERSION = "api-version=2017-12-01";
+    private static AIMS_FORMAT = "format=json";
+
+    private _collectionInterval: number = 900000;
+    private _client: TelemetryClient;
+    private _handle: NodeJS.Timer;
+    private _isEnabled: boolean;
+    private _isInitialized: boolean;
+    private _isVM: boolean = false;
+    private _vmData = <{[key: string]: string}>{};
+
+    constructor(client: TelemetryClient) {
+        if (!HeartBeat.INSTANCE) {
+            HeartBeat.INSTANCE = this;
+        }
+
+        this._isInitialized = false;
+        this._client = client;
+    }
+
+    public enable(isEnabled: boolean, config?: Config) {
+        this._isEnabled = isEnabled;
+        if (this._isEnabled && !this._isInitialized) {
+            this._isInitialized = true;
+        }
+
+        if (isEnabled) {
+            if (!this._handle) {
+                this._handle = setInterval(() => this.trackHeartBeat(config, () => {}), this._collectionInterval);
+                this._handle.unref(); // Allow the app to terminate even while this loop is going on
+            }
+        } else {
+            if (this._handle) {
+                clearInterval(this._handle);
+                this._handle = undefined;
+            }
+        }
+    }
+
+    public isInitialized() {
+        return this._isInitialized;
+    }
+
+    public static isEnabled() {
+        return HeartBeat.INSTANCE && HeartBeat.INSTANCE._isEnabled;
+    }
+
+    public trackHeartBeat(config: Config, callback: () => void) {
```

Comment:
hi @markwolff , is this the callback you mentioned? @MSNev helped me work this out here. Do you see any issues? The promisify from util is "undefined" here.

Reply:
Yes this looks good